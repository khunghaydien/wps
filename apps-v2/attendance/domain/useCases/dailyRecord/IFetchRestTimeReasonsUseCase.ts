import { RestTimeReason } from '@attendance/domain/models/RestTimeReason';

import * as Interface from '@attendance/domain/useCases/IUseCase';

export type IInputData = Interface.IInputData<{
  employeeId?: string | null;
  targetDate: string;
}>;

export type IOutputData = Interface.IOutputData<{
  employeeId: string;
  targetDate: string;
  restReasons: RestTimeReason[];
}>;

export type IUseCase = Interface.IUseCase<IInputData, IOutputData>;
