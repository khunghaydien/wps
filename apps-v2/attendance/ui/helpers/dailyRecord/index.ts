import * as restTime from './restTime';
import * as time from './time';

export { restTime, time };
