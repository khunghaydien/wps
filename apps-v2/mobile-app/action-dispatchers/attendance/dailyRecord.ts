import isNil from 'lodash/isNil';
import moment from 'moment';

import msg from '../../../commons/languages';
import { showAlert } from '../../modules/commons/alert';

import { Status as DailyRequestStatus } from '@attendance/domain/models/AttDailyRequest';
import {
  getAttDailyRecordByDate,
  isTargetDateInTimesheet,
  Timesheet,
} from '@attendance/domain/models/Timesheet';
import { WorkingType } from '@attendance/domain/models/WorkingType';

import { State as UiDailyEditingState } from '../../modules/attendance/timesheet/ui/daily/editing';
import { actions as EntitiesApprovalHistoriesActions } from '@mobile/modules/attendance/dailyRequest/entities/approvalHistories';

import * as RecordsUtil from '@attendance/libraries/utils/Records';

import {
  clearTimesheet,
  setTimesheet,
} from '@mobile/containers/pages/attendance/TimesheetDailyPageContainer/presenters/timesheet/fetch';
import UseCases from '@mobile/containers/pages/attendance/TimesheetDailyPageContainer/UseCases';

import { AppDispatch } from '../AppThunk';

const getTodayData = (targetDate: string, timesheet: Timesheet) => {
  const workingType = RecordsUtil.getWithinRange(
    targetDate,
    timesheet.workingTypes
  );
  const record = getAttDailyRecordByDate(targetDate, timesheet);
  return {
    workingType,
    record,
  };
};

const resetTimesheet = (targetDate: string): Promise<Timesheet | null> =>
  UseCases()
    .fetchTimesheet({
      targetDate,
    })
    .then(({ timesheet }) => timesheet)
    .catch(() => null);

const convertToDailyRecord = (
  editing: UiDailyEditingState,
  workingType: WorkingType
) => ({
  recordId: editing.id,
  recordDate: editing.targetDate,
  startTime: editing.startTime,
  endTime: editing.endTime,
  restTimes: editing.restTimes,
  restHours: editing.restHours,
  otherRestReason: editing.otherRestReason,
  commuteCount: workingType?.useManageCommuteCount
    ? editing.commuteCount
    : null,
  remarks: editing.remarks || '',
});

const fetchApprovalRequestHistoryRepository =
  (targetDate: string, timesheet: Timesheet) =>
  async (dispatch: AppDispatch) => {
    const { workingType, record } = getTodayData(targetDate, timesheet);
    const fixDailyRequest = record?.fixDailyRequest;

    if (workingType?.useFixDailyRequest && fixDailyRequest?.id) {
      await UseCases().fetchApprovalRequestHistory(fixDailyRequest.id);
    } else {
      dispatch(EntitiesApprovalHistoriesActions.initialize([]));
    }
  };

const fetchRestTimeReasons = async (
  targetDate: string,
  timesheet: Timesheet
) => {
  const { workingType, record } = getTodayData(targetDate, timesheet);

  if (workingType?.useRestReason && record?.editable) {
    return UseCases().fetchRestTimeReasons({
      targetDate,
    });
  }
};

/**
 * Timesheet を取得する
 * 勤務表で既に Timesheet を取得しているはずなので cache があれば cache を使用する。
 * @param targetDate
 * @param timesheet
 * @returns
 */
const initializeTimesheet =
  (targetDate: string, timesheet?: Timesheet) => (dispatch: AppDispatch) => {
    if (!isNil(timesheet) && isTargetDateInTimesheet(timesheet, targetDate)) {
      dispatch(setTimesheet(targetDate, timesheet));
      return timesheet;
    } else {
      return resetTimesheet(targetDate);
    }
  };

/**
 * 初期化処理
 * @param targetDate
 * @param timesheet
 * @returns
 */
export const initialize =
  (targetDate?: string, timesheet?: Timesheet) =>
  async (dispatch: AppDispatch): Promise<void> => {
    dispatch(clearTimesheet());
    const $targetDate = targetDate || moment().format('YYYY-MM-DD');

    const [_, $timesheet] = await Promise.all([
      UseCases().fetchUserSetting(),
      dispatch(initializeTimesheet($targetDate, timesheet)),
    ]);

    if ($timesheet && $timesheet.isMigratedSummary) {
      dispatch(showAlert(msg().Att_Err_CanNotDisplayBeforeUsing));
    }

    // 休憩理由と承認履歴は日毎にローディングが必要なので毎回取得する
    await Promise.all([
      fetchRestTimeReasons($targetDate, $timesheet),
      dispatch(fetchApprovalRequestHistoryRepository($targetDate, $timesheet)),
    ]);
  };

export const saveDailyRecord = async (
  editing: UiDailyEditingState,
  workingType: WorkingType
) => {
  const { targetDate } = editing;
  await UseCases().saveDailyRecord(convertToDailyRecord(editing, workingType));
  await resetTimesheet(targetDate);
};

export const fixDailyRequest =
  (
    record: UiDailyEditingState,
    remarkableStatus: DailyRequestStatus,
    workingType: WorkingType
  ) =>
  async (dispatch: AppDispatch) => {
    const { targetDate } = record;

    const result = await UseCases().submitFixDailyRequestAndSaveDailyRecord({
      dailyRecord: convertToDailyRecord(record, workingType),
      dailyRequestSummary: { status: remarkableStatus },
    });

    // 保存処理は複数の API が呼ばれていている。
    // どれか一つが成功しているかもしれないので必ず Timesheet は更新する。
    const timesheet = await resetTimesheet(targetDate);

    if (result) {
      // 失敗した場合は承認履歴の呼び出しは必要ない
      await dispatch(
        fetchApprovalRequestHistoryRepository(targetDate, timesheet)
      );
    }
  };

export const cancelApprovalDailyRequest =
  (record: UiDailyEditingState) => async (dispatch: AppDispatch) => {
    const { targetDate } = record;
    const { result } = await UseCases().cancelApprovalFixDailyRequest(
      record.fixDailyRequest.id
    );
    if (!result) {
      return;
    }
    const timesheet = await resetTimesheet(targetDate);
    await dispatch(
      fetchApprovalRequestHistoryRepository(targetDate, timesheet)
    );
  };

export const cancelDailyRequest =
  (record: UiDailyEditingState) => async (dispatch: AppDispatch) => {
    const { targetDate } = record;
    const { result } = await UseCases().cancelSubmittedFixDailyRequest(
      record.fixDailyRequest.id
    );
    if (!result) {
      return;
    }
    const timesheet = await resetTimesheet(targetDate);
    await dispatch(
      fetchApprovalRequestHistoryRepository(targetDate, timesheet)
    );
  };
