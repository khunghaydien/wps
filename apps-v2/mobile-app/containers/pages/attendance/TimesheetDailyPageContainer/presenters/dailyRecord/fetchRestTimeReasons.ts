import { Store } from 'redux';

import { AppDispatch } from '@mobile/modules/commons/AppThunk';
import { catchApiError } from '@mobile/modules/commons/error';
import { endLoading, startLoading } from '@mobile/modules/commons/loading';

import {
  clear,
  set,
} from '@mobile/modules/attendance/timesheet/ui/daily/restReasons';

import { IPresenter } from '@attendance/application/useCaseInteractors/dailyRecord/FetchRestTimeReasonsUseCaseInteractor';

export default ({ dispatch }: Store) =>
  (): IPresenter => {
    let loadingId;
    return {
      start: () => {
        loadingId = (dispatch as AppDispatch)(startLoading());
      },
      complete: ({ restReasons }) => {
        dispatch(set(restReasons));
      },
      error: (err) => {
        dispatch(clear());
        dispatch(catchApiError(err as Parameters<typeof catchApiError>[0]));
      },
      finally: () => {
        (dispatch as AppDispatch)(endLoading(loadingId));
      },
    };
  };
