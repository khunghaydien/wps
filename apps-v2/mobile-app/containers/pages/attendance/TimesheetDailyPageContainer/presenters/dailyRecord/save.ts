import { Store } from 'redux';

import msg from '@commons/languages';
import { showToast } from '@commons/modules/toast';
import TextUtil from '@commons/utils/TextUtil';
import { AppDispatch } from '@mobile/modules/commons/AppThunk';
import { showConfirm } from '@mobile/modules/commons/confirm';
import { catchApiError } from '@mobile/modules/commons/error';
import { endLoading, startLoading } from '@mobile/modules/commons/loading';

import { IPresenter } from '@attendance/application/useCaseInteractors/dailyRecord/SaveUseCaseInteractor';
import { bind } from '@attendance/libraries/Collection';

const normal = (store: Store) => (): IPresenter => {
  let loadingId;
  const dispatch = store.dispatch as AppDispatch;
  return {
    start: () => {
      loadingId = dispatch(startLoading());
    },
    complete: () => {
      dispatch(showToast(msg().Att_Lbl_SaveWorkTime));
    },
    error: (err) => {
      if (err instanceof Array) {
        // より強力なエラーは reducer 内で上書きされないようになっているため
        // すべてのエラーを dispatch しておく。
        // 後勝ちになってしまうので reverse して実質的に先勝ちにする。
        err.reverse().forEach((e) => {
          dispatch(catchApiError(e));
        });
      } else {
        dispatch(catchApiError(err as Parameters<typeof catchApiError>[0]));
      }
    },
    confirmToComplementInsufficientingRestTime: async ({
      insufficientRestTime,
    }) => {
      dispatch(endLoading(loadingId));

      const confirmText = TextUtil.template(
        msg().Com_Msg_InsufficientRestTime,
        insufficientRestTime
      );
      const result = await dispatch(showConfirm(confirmText));
      loadingId = dispatch(startLoading());
      return result;
    },
    finally: () => {
      dispatch(endLoading(loadingId));
    },
  };
};

const noCompleteMessage = (store: Store) => () => ({
  ...normal(store)(),
  complete: () => {},
});

export default (store: Store) =>
  bind(
    {
      normal,
      noCompleteMessage,
    },
    store
  );
