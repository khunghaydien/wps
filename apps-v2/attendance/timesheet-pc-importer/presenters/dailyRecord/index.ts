import fetchRestTimeReasons from './fetchRestTimeReasons';
import fetchRestTimeReasonsForBulk from './fetchRestTimeReasonsForBulk';
import { bind } from '@attendance/libraries/Collection';
import { AppStore } from '@attendance/timesheet-pc-importer/store/AppStore';

export default (store: AppStore) =>
  bind({ fetchRestTimeReasons, fetchRestTimeReasonsForBulk }, store);
