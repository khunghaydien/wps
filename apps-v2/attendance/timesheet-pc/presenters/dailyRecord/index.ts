import { Store } from 'redux';

import fetchRestTimeReasons from './fetchRestTimeReasons';
import save from './save';
import saveFields from './saveFields';
import { bind } from '@attendance/libraries/Collection';

export default (store: Store) =>
  bind(
    {
      save,
      fetchRestTimeReasons,
      saveFields,
    },
    store
  );
