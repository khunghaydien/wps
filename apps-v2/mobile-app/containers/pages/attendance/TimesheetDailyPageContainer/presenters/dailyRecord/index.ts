import { Store } from 'redux';

import fetchRestTimeReasons from './fetchRestTimeReasons';
import save from './save';
import { bind } from '@attendance/libraries/Collection';

export default (store: Store) =>
  bind(
    {
      save,
      fetchRestTimeReasons,
    },
    store
  );
