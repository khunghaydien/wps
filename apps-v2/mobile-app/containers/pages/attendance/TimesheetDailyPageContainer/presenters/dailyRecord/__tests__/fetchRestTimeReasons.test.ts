import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import uuid from 'uuid/v4';

import Presenter from '../fetchRestTimeReasons';
import { IOutputData } from '@attendance/domain/useCases/dailyRecord/IFetchRestTimeReasonsUseCase';

jest.mock('uuid/v4');

const mockStore = configureMockStore([thunk]);

beforeEach(() => {
  jest.clearAllMocks();

  let i = 1;
  (uuid as jest.Mock).mockImplementation(() => `${i++}`);
});

describe('complete()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore({});
    const presenter = Presenter(store)();

    // Act
    presenter.complete({
      restReasons: 'restReasons',
    } as unknown as IOutputData);

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('error()', () => {
  it('should do', () => {
    // Arrange
    const store = mockStore({});
    const presenter = Presenter(store)();

    // Act
    presenter.error([
      {
        message: 'Error Test 1',
      },
      {
        message: 'Error Test 2',
      },
    ]);

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
  it('should do if Error is Array', () => {
    // Arrange
    const store = mockStore({});
    const presenter = Presenter(store)();

    // Act
    presenter.error({
      message: 'Error Test',
    });

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('start()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore({});
    const presenter = Presenter(store)();

    // Act
    presenter.start();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('finally()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore({});
    const presenter = Presenter(store)();

    // Act
    presenter.finally();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

it('should do with loadingId.', () => {
  // Arrange
  const store = mockStore({});
  const presenter = Presenter(store)();

  // Act
  presenter.start();
  presenter.finally();

  // Assert
  expect(store.getActions()).toMatchSnapshot();
});
