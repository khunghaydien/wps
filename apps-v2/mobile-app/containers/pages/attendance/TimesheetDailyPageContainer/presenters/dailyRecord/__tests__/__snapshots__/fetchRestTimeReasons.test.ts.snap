// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complete() should do. 1`] = `
Array [
  Object {
    "payload": "restReasons",
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/REST_REASONS/SET",
  },
]
`;

exports[`error() should do 1`] = `
Array [
  Object {
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/REST_REASONS/CLEAR",
  },
  Object {
    "payload": Object {
      "errorCode": undefined,
      "isContinuable": true,
      "isFunctionCantUseError": false,
      "message": "APIエラー - undefined",
      "problem": undefined,
      "stackTrace": "",
      "type": "APIエラー",
    },
    "type": "CATCH_API_ERROR",
  },
]
`;

exports[`error() should do if Error is Array 1`] = `
Array [
  Object {
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/REST_REASONS/CLEAR",
  },
  Object {
    "payload": Object {
      "errorCode": undefined,
      "isContinuable": true,
      "isFunctionCantUseError": false,
      "message": "APIエラー - Error Test",
      "problem": "Error Test",
      "stackTrace": "",
      "type": "APIエラー",
    },
    "type": "CATCH_API_ERROR",
  },
]
`;

exports[`finally() should do. 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "END_LOADING",
  },
]
`;

exports[`should do with loadingId. 1`] = `
Array [
  Object {
    "payload": "1",
    "type": "START_LOADING",
  },
  Object {
    "payload": "1",
    "type": "END_LOADING",
  },
]
`;

exports[`start() should do. 1`] = `
Array [
  Object {
    "payload": "1",
    "type": "START_LOADING",
  },
]
`;
